speaker/bgp_controller.go

```diff
@@ -71,14 +87,16 @@ newPeers:
 		if p == nil {
 			continue
 		}
-		l.Log("event", "peerRemoved", "peer", p.cfg.Addr, "reason", "removedFromConfig", "msg", "peer deconfigured, closing BGP session")
-		if p.bgp != nil {
-			if err := p.bgp.Close(); err != nil {
-				l.Log("op", "setConfig", "error", err, "peer", p.cfg.Addr, "msg", "failed to shut down BGP session")
+		l.Log("event", "peerDeconfigured", "peer", p.Cfg.Addr, "reason", "removedFromConfig", "msg", "peer deconfigured, closing BGP session")
+		if p.BGP != nil {
+			if err := p.BGP.Close(); err != nil {
+				l.Log("op", "setConfig", "error", err, "peer", p.Cfg.Addr, "msg", "failed to shut down BGP session")
 			}
 		}
 	}
 
+	c.discoverNodePeer(l)
```

Comment:
I'm having a hard time trying to see if we can change this to something way simpler. It might not be possible, so take this with a grain of salt.

I think the patch would be way reduced and simplified if here we can just discover the peer and add it the the peer slice (`c.peers`). My main idea is just that.

I think we talked about it online, but when we talked about you don't remember the reason and I wonder if, given the number of iterations this patch has, those might not apply anymore :crossed_fingers: 

One way I'm thinking to achieve that is:
 * Make `c.discoverNodePeer(l)` return a peer (probably rename the function too). This function will do the same it does today, but just return the peer at the end. IOW, instead of [setting c.NodePeer][discoverNodePeer] return it.
 * Add that peer to the peer slice. We might need to check for duplicates and remove one. But it is the same "which one" we remove (the one that come from the configmap or from `discoverNodePeer()`: they are equal. Later, if a config or label change (changing the peers), this whole logic will be executed and if now the peers happen to be different, both will be in the peers slice.
 * Change `SetNode()` in this file to call `SetConfig()` so it discovers the peers and sets the config, if needed.

With that, IIUC, then we won't need any modifications to the rest of the functions (like `syncBGPSessions`) while making this semantically way simpler: just the functions to set the peers need to be aware of the PeerAutodiscovery thingy, all the other functions are the same as they operate on a slice of peers.

As another side effect is that the change is way safer too. We mostly need to parse the peer and add it with care to the slice. The rest is unchanged.

Maybe this makes sense on my head only, though. What do you think?

[discoverNodePeer]: https://github.com/metallb/metallb/pull/593/commits/a7c1d8b803248f393e09e6dd322e9d9a66c8a7cd#diff-83ec95a1dc9f40aea0f9469e2acc035eR298

Reply:
I've started with your suggested approach when I started working on this feature. I had a hard time getting something which works this way. Unfortunately I can't recall exactly which walls I hit when attempting this. Do you think it's worth investing more time in re-attempting this given that we already have a working and tested implementation?

Reply:
Yes, my guess is what was missing was the current config, with the current config it should be possible to just discover the peer and delete old ones.

IMHO, we already connect to an arbitrary number of peers. It makes sense to just add a peer there and not touch `syncPeers()`, `syncBGPSessions()` with special cases. IMHO, it would be nice (and I think is possible as I described) to change it and simplify it.